File: glyphs_generator/generate.py
from functools import reduce
from typing import Dict, List

from .data import Glyph, InternalGlyph, InternalStroke, Stroke, GeneratorParameters
from .intersect import are_strokes_linked


class GlyphGenerator:
    def __init__(self, parameters: GeneratorParameters):
        self.parent_strokes = parameters.parent_strokes
        self.intersection_matrix = parameters.intersection_matrix
        self.transformation_table = parameters.transformation_matrix
        self.nb_transformations = len(parameters.transformation_matrix[0])

    def are_strokes_intersecting(self, glyph: InternalGlyph) -> bool:
        indices = [stroke.index for stroke in glyph.strokes]
        return are_strokes_linked(indices, self.intersection_matrix)

    def transform(self, glyph: InternalGlyph) -> List[InternalGlyph]:
        result = [InternalGlyph.empty() for _ in range(self.nb_transformations)]
        for i in range(self.nb_transformations):
            for stroke in glyph.strokes:
                stroke_index = self.transformation_table[stroke.index][i]
                glyph_ = InternalGlyph.from_stroke(InternalStroke(index=stroke_index))
                result[i] = result[i] | glyph_
        return result

    def to_glyph(self, glyph: InternalGlyph) -> Glyph:
        return Glyph(strokes=[self.parent_strokes[s.index] for s in glyph.strokes])

    def from_stroke(self, stroke: Stroke) -> InternalGlyph:
        for i, stroke_ in enumerate(self.parent_strokes):
            if (
                stroke_.x0 == stroke.x0
                and stroke_.x1 == stroke.x1
                and stroke_.y0 == stroke.y0
                and stroke_.y1 == stroke.y1
            ) or (
                stroke_.x0 == stroke.x1
                and stroke_.x1 == stroke.x0
                and stroke_.y0 == stroke.y1
                and stroke_.y1 == stroke.y0
            ):
                return InternalGlyph(strokes=[InternalStroke(index=i)], identifier=2**i)
        raise Exception(f"Could not find stroke {stroke} in parent strokes")

    def generate(self, strokes: List[Stroke], seed: Stroke) -> List[Glyph]:
        n = len(strokes)
        seed_internal = self.from_stroke(seed)
        strokes_internal = [self.from_stroke(stroke) for stroke in strokes]
        glyphs: Dict[int, List[InternalGlyph]] = {i + 1: [] for i in range(n)}
        glyphs[1].append(seed_internal)
        for i, last_glyphs in glyphs.items():
            if i == n:
                break
            for glyph in last_glyphs:
                for stroke in strokes_internal:
                    next_glyph = glyph | stroke
                    if (
                        len(next_glyph.strokes) == i + 1
                        and next_glyph not in glyphs[i + 1]
                        and self.are_strokes_intersecting(next_glyph)
                    ):
                        transformed_glyphs = self.transform(next_glyph)
                        if all([g not in glyphs[i + 1] for g in transformed_glyphs]):
                            glyphs[i + 1].append(next_glyph)
        return list(map(lambda x: self.to_glyph(x), reduce(lambda x, y: x + y, glyphs.values())))

File: src/generate.rs
use crate::glyph::Glyph;
use crate::glyph::InternalGlyph;
use crate::parameters::Parameters;
use crate::stroke::InternalStroke;
use crate::stroke::Stroke;

#[derive(Debug, Clone)]
pub struct GlyphGenerator {
    pub parameters: Parameters,
}

impl GlyphGenerator {
    pub fn new(parameters: Parameters) -> Self {
        Self { parameters }
    }

    pub fn are_strokes_intersecting(&self, glyph: &InternalGlyph) -> bool {
        glyph.strokes.iter().all(|stroke| {
            glyph.strokes.iter().any(|other| {
                stroke.index != other.index
                    && self.parameters.intersection_matrix[stroke.index][other.index] == 1
            })
        })
    }

    pub fn transform(&self, glyph: &InternalGlyph) -> Vec<InternalGlyph> {
        let nb_transformations = self.parameters.transformation_matrix[0].len();
        (0..nb_transformations)
            .map(|i| {
                glyph
                    .strokes
                    .iter()
                    .fold(InternalGlyph::empty(), |acc, stroke| {
                        let stroke_index = self.parameters.transformation_matrix[stroke.index][i];
                        acc.union(&InternalGlyph::from_stroke(InternalStroke {
                            index: stroke_index,
                        }))
                    })
            })
            .collect()
    }

    pub fn to_glyph(&self, glyph: &InternalGlyph) -> Glyph {
        Glyph {
            strokes: glyph
                .strokes
                .iter()
                .map(|s| self.parameters.parent_strokes[s.index].clone())
                .collect(),
        }
    }

    pub fn from_stroke(&self, stroke: &Stroke) -> InternalGlyph {
        self.parameters
            .parent_strokes
            .iter()
            .enumerate()
            .find_map(|(i, parent_stroke)| {
                if stroke == parent_stroke || stroke.is_inverse(parent_stroke) {
                    Some(InternalGlyph::from_stroke(InternalStroke { index: i }))
                } else {
                    None
                }
            })
            .expect("Could not find stroke in parent strokes")
    }

    pub fn generate(&self, strokes: &[Stroke], seed: &Stroke) -> Vec<Glyph> {
        let n = strokes.len();
        let seed_internal = self.from_stroke(seed);
        let strokes_internal: Vec<InternalGlyph> = strokes
            .iter()
            .map(|stroke| self.from_stroke(stroke))
            .collect();

        let mut glyphs: Vec<Vec<InternalGlyph>> = vec![Vec::new(); n];
        glyphs[0].push(seed_internal);

        for i in 0..n - 1 {
            let mut next_glyphs = Vec::new();

            for glyph in &glyphs[i] {
                for stroke in &strokes_internal {
                    let next_glyph = glyph.union(stroke);
                    if next_glyph.strokes.len() < i + 2 {
                        continue;
                    }
                    if next_glyphs.contains(&next_glyph) {
                        continue;
                    }
                    if self.are_strokes_intersecting(&next_glyph) {
                        let transformed_glyphs = self.transform(&next_glyph);
                        if transformed_glyphs
                            .iter()
                            .all(|g| !glyphs[i + 1].contains(&g))
                        {
                            next_glyphs.push(next_glyph);
                        }
                    }
                }
            }

            glyphs[i + 1].extend(next_glyphs);
        }

        // Convert and flatten the glyphs to the desired type
        glyphs
            .into_iter()
            .flatten()
            .map(|g| self.to_glyph(&g))
            .collect()
    }
}
